### PR TITLE
Support UPN and SID.

### PR DESCRIPTION
**Pull request by anhu**
This is specifically for support in the compatibility layer. 

Fixes ZD15759

# Testing

Added unit tests. 

# Checklist

 - [x] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

**Comment by anhu** [comment]
Jenkins retest this please

File: tests/api.c [hunk 1]
```diff
@@ -43264,6 +43264,142 @@ static int test_wolfSSL_X509_NAME_ENTRY(void)
     return res;
 }
 
+/* Note the lack of wolfSSL_ prefix...this is a compatability layer test. */
+static int test_GENERAL_NAME_set0_othername(void) {
+    int res = TEST_SKIPPED;
+#if defined(OPENSSL_EXTRA) && !defined(NO_CERTS) && \
+    defined(WOLFSSL_CERT_GEN) && defined(WOLFSSL_CERT_REQ) && \
+    defined(WOLFSSL_CUSTOM_OID) && defined(WOLFSSL_ALT_NAMES) && \
```
Comment (JacobBarthelmeh) [inline tests/api.c:43272]: Am not seeing WOLFSSL_CUSTOM_OID in configure.ac or in the testing flags for CI. Did I miss where the macro is being defined somewhere and this test being ran with CI testing.

File: wolfssl/ssl.h [hunk 2]
```diff
@@ -1557,6 +1557,11 @@ WOLFSSL_API int wolfSSL_GENERAL_NAME_set_type(WOLFSSL_GENERAL_NAME* name,
         int typ);
 WOLFSSL_API WOLFSSL_GENERAL_NAMES* wolfSSL_GENERAL_NAMES_dup(
                                              WOLFSSL_GENERAL_NAMES* gns);
+WOLFSSL_API int wolfSSL_GENERAL_NAME_set0_othername(WOLFSSL_GENERAL_NAME* gen,
+                                                    WOLFSSL_ASN1_OBJECT* oid,
+                                                    WOLFSSL_ASN1_TYPE* value);
+
+WOLFSSL_API WOLFSSL_STACK* wolfSSL_sk_GENERAL_NAME_new(void *cmp_func);
```
Comment (JacobBarthelmeh) [inline wolfssl/ssl.h:1564]: Why cmp_func for the parameter name, is it a compare function? Lets also use camel case here.

File: wolfssl/wolfcrypt/asn_public.h [hunk 3]
```diff
@@ -421,22 +419,22 @@ typedef struct CertName {
...
-#endif
 
 /* for user to fill for certificate generation */
 typedef struct Cert {
     int      version;                   /* x509 version  */
     byte     serial[CTC_SERIAL_SIZE];   /* serial number */
     int      serialSz;                  /* serial size */
     int      sigType;                   /* signature algo type */
+#if defined(WOLFSSL_CERT_GEN) || defined(OPENSSL_EXTRA) \
+ || defined(OPENSSL_EXTRA_X509_SMALL)
```
Comment (JacobBarthelmeh) [inline wolfssl/wolfcrypt/asn_public.h:432]: it's better to avoid variable size struct here and try to keep it constant for all builds

**Review by JacobBarthelmeh — CHANGES_REQUESTED** [review]

File: wolfssl/ssl.h (hunk 2 above)
Comment (anhu) [reply wolfssl/ssl.h:1564]: Can't use camel case here.  Following OpenSSL naming conventions.

File: wolfssl/wolfcrypt/asn_public.h (hunk 3 above)
Comment (anhu) [reply wolfssl/wolfcrypt/asn_public.h:432]: I'm with you on that one!!  Thanks!

File: wolfssl/ssl.h (hunk 2 above)
Comment (JacobBarthelmeh) [reply wolfssl/ssl.h:1564]: For the name of the parameter in the function signature it should not matter? Mapping from OpenSSL -> wolfSSL should happen in the wolfssl/openssl/* files, does it throw an error or break when changed?

Comment (anhu) [reply wolfssl/ssl.h:1564]: Sorry, I thought convention was to use what openSSL used. I will change to `cmpFunc`.

File: src/x509.c [hunk 4]
```diff
@@ -9228,6 +9466,21 @@ WOLF_STACK_OF(WOLFSSL_X509)* wolfSSL_X509_chain_up_ref(
                     }
                 }
             }
+
+        #ifdef WOLFSSL_CUSTOM_OID
+            if ((req->customExtCount < 0) ||
+                (req->customExtCount >= NUM_CUSTOM_EXT)) {
+                WOLFSSL_MSG("Bad value for customExtCount.");
+                ret = WOLFSSL_FAILURE;
```
Comment (JacobBarthelmeh) [inline src/x509.c:9474]: Setting ret here does not avoid the upcoming for loop and accessing cert->customCertExt.

File: wolfssl/wolfcrypt/asn_public.h (hunk 3 above)
Comment (JacobBarthelmeh) [reply wolfssl/wolfcrypt/asn_public.h:432]: can we change this to not be a dynamic struct. The reason being is this Cert struct for older builds was dereferenced directly

File: tests/api.c (hunk 1 above)
Comment (JacobBarthelmeh) [reply tests/api.c:43272]: maybe not in this PR but please either a) add WOLFSSL_CUSTOM_OID to --enable-all or b) add a specific WOLFSSL_CUSTOM_OID test to our CI enable flag list

File: src/x509.c (hunk 4 above)
Comment (anhu) [reply src/x509.c:9474]: Good catch!!

**Comment by anhu** [comment]
Did a squash and rebase to see if it would fix the failing tests which feel like they might not be related to my changes.

File: wolfssl/wolfcrypt/asn_public.h (hunk 3 above)
Comment (anhu) [reply wolfssl/wolfcrypt/asn_public.h:433]: Tried to remove it but gets CertName unknown type. :( 

also, there are several other ones in there already: 


```#ifdef WOLFSSL_ALT_NAMES
    byte     altNames[CTC_MAX_ALT_SIZE]; /* altNames copy */
    int      altNamesSz;                 /* altNames size in bytes */
#endif
    byte     beforeDate[CTC_DATE_SIZE];  /* before date copy */
    int      beforeDateSz;               /* size of copy */
    byte     afterDate[CTC_DATE_SIZE];   /* after date copy */
    int      afterDateSz;                /* size of copy */
#ifdef WOLFSSL_CERT_EXT
    byte    skid[CTC_MAX_SKID_SIZE];     /* Subject Key Identifier */
    int     skidSz;                      /* SKID size in bytes */
    byte    akid[CTC_MAX_AKID_SIZE
#ifdef WOLFSSL_AKID_NAME
              + sizeof(CertName) + CTC_SERIAL_SIZE
#endif
              ]; /* Authority Key
                                                         * Identifier */
    int     akidSz;                      /* AKID size in bytes */
#ifdef WOLFSSL_AKID_NAME
    byte    rawAkid;                     /* Set to true if akid is a
                                          * AuthorityKeyIdentifier object.
                                          * Set to false if akid is just a
                                          * KeyIdentifier object. */
#endif
```

**Review by JacobBarthelmeh — APPROVED** [review]

**Pull request merged**